### PR TITLE
version 1.0.0 : initial ledger firmware release 

### PR DESCRIPTION
**Pull request by coolcottontail**
This PR contains the following:

###  initial implementation of ledger firmware meeting ledger's requirement:

- [x] Verify public address on the Ledger device	
- [x] Display transaction information before allowing signature	
- [x] Reject a transaction on the Ledger device	
- [x] Sign a transaction on the Ledger device

### simple companion test application to send commands to ledger firmware

These two are required by ledger security review.

### CODE REVIEW

File: src/getPublicKey.c [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        case 2:
+            return (ctx->displayIndex >= 42 - 12) ? NULL : element;
+        default:
+            return element;
+    }
+}
+
+// Define the button handler for the comparison screen. Again, this is nearly
+// identical to the signHash comparison button handler.
+static unsigned int ui_getPublicKey_compare_button(unsigned int button_mask, unsigned int button_mask_counter) {
```
Comment (fxfactorial) [inline src/getPublicKey.c]: Why all these are `unsigned int` and not `size_t`

File: src/getTxnSig.c [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+        UI_ICON_LEFT(0x01, BAGL_GLYPH_ICON_LEFT),
+        UI_ICON_RIGHT(0x02, BAGL_GLYPH_ICON_RIGHT),
+
+        UI_TEXT(0x00, 0, 12, 128, "Send to Addr:"),
+        UI_TEXT(0x00, 0, 26, 128, global.signTxnContext.partialAddrStr),
+};
+
+
+static const bagl_element_t* ui_prepro_address_compare(const bagl_element_t *element) {
+    switch (element->component.userid) {
```
Comment (fxfactorial) [inline src/getTxnSig.c]: What is the invariant that `element` is never NULL

File: src/getTxnSig.c [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+                ctx->displayIndex--;
+            }
+
+            os_memmove(ctx->partialAddrStr, ctx->toAddr+ctx->displayIndex, 12);
+            // Re-render the screen.
+            UX_REDISPLAY();
+            break;
+
+        case BUTTON_RIGHT:
+        case BUTTON_EVT_FAST | BUTTON_RIGHT: // SEEK RIGHT
```
Comment (fxfactorial) [inline src/getTxnSig.c]: Why `BUTTON_LEFT`, `BUTTON_RIGHT` not directly folded into their following case, like you did with `  case BUTTON_EVT_RELEASED | BUTTON_LEFT | BUTTON_RIGHT:` Is it for future code explicit for button_left, button_right?

File: src/harmony.c [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+
+void deriveOneKeypair(cx_ecfp_private_key_t *privateKey, cx_ecfp_public_key_t *publicKey) {
+	uint8_t keySeed[32];
+	cx_ecfp_private_key_t pk;
+
+	// bip32 path for ONE is 44'/1023'/0'/0/0
+	uint32_t bip32Path[] = {44 | 0x80000000, 1023 | 0x80000000,  0x80000000, 0, 0};
+    os_perso_derive_node_bip32(CX_CURVE_256K1, bip32Path, 5, keySeed, NULL);
+    cx_ecfp_init_private_key(CX_CURVE_256K1, keySeed, 32, &pk);
```
Comment (fxfactorial) [inline src/harmony.c]: Why are we doing this work in case when both privateKey, publicKey are both NULL

File: src/harmony.c [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+		dst[2*i+0] = hex[(data[i]>>4) & 0x0F];
+		dst[2*i+1] = hex[(data[i]>>0) & 0x0F];
+	}
+	dst[2*inlen] = '\0';
+}
+
+void getEthAddressFromKey(cx_ecfp_public_key_t *publicKey, uint8_t *out) {
+    cx_sha3_t sha3Context;
+    uint8_t hashAddress[32];
+    cx_keccak_init(&sha3Context, 256);
```
Comment (fxfactorial) [inline src/harmony.c]: Does this need a fresh `init` on the sha3Context each time its used?

File: src/main.c [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+handler_fn_t handleGetVersion;
+handler_fn_t handleGetPublicKey;
+//handler_fn_t handleSignHash;
+handler_fn_t handleSignTx;
+
+static handler_fn_t* lookupHandler(uint8_t ins) {
+	switch (ins) {
+	case INS_GET_VERSION:    return handleGetVersion;
+	case INS_GET_PUBLIC_KEY: return handleGetPublicKey;
+	//case INS_SIGN_HASH:      return handleSignHash;
```
Comment (fxfactorial) [inline src/main.c]: why signHash disabled

Comment (coolcottontail) [reply src/main.c]: Disabled just for security reason. As there will be no display of tx related information in ledger Nano S. if anyone hacked the client side app , they can send any hash (from any tx) to ledger Nano S to sign.

File: README.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+  src/           firmware source code for Ledger Nano S
+  wallet/        harmony-one hardware wallet main application code
+  icos/          gif files for Ledger Nano S app icon and Ledger Live icon
+  glyphs/        icon files used in firmware GUI
+```
+
+## Build Environment
+
+### Docker toolchain image
+In order to make compiling as eas as possible you can make use of a docker image containing all the necessary compilers and the [nanos-secure-sdk](https://github.com/LedgerHQ/nanos-secure-sdk).
```
Comment (LeoHChen) [inline README.md]: eas?

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,21 @@
+FROM zondax/ledger-docker-bolos
```
Comment (LeoHChen) [inline Dockerfile]: are we building our own docker image?

File: oneledger.go [hunk 9]
```diff
@@ -0,0 +1,483 @@
...
+		}
+		oneAddr, err := nanos.GetAddress()
+		if err != nil {
+        	log.Fatalln("Couldn't get one address:", err)
+        }
+		fmt.Println(oneAddr)
+
+	case signHashCmd:
+	    fmt.Println("Sign hash is disabled in ledger firmware, please use signtx instead.")
+	/*
```
Comment (LeoHChen) [inline oneledger.go]: remove the commented code please.

File: Dockerfile (hunk 8 above)
Comment (LeoHChen) [reply Dockerfile]: is there a document on how to publish our own docker image?

File: src/base32.c [hunk 10]
```diff
@@ -0,0 +1,80 @@
+/*
```
Comment (LeoHChen) [inline src/base32.c]: can you provide some reference of this code? where is it coming from?

File: src/getPublicKey.c [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+	UI_ICON_LEFT(0x01, BAGL_GLYPH_ICON_LEFT),
+	UI_ICON_RIGHT(0x02, BAGL_GLYPH_ICON_RIGHT),
+	UI_TEXT(0x00, 0, 12, 128, "Compare:"),
+	// The visible portion of the public key or address.
+	UI_TEXT(0x00, 0, 26, 128, global.getPublicKeyContext.partialStr),
+};
+
+static const bagl_element_t* ui_prepro_getPublicKey_compare(const bagl_element_t *element) {
+    switch (element->component.userid) {
+        case 1:
```
Comment (LeoHChen) [inline src/getPublicKey.c]: what's these 1 and 2? any comments?

File: src/getPublicKey.c [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+		if (ctx->displayIndex > 0) {
+			ctx->displayIndex--;
+		}
+		os_memmove(ctx->partialStr, ctx->fullStr+ctx->displayIndex, 12);
+		UX_REDISPLAY();
+		break;
+
+	case BUTTON_RIGHT:
+	case BUTTON_EVT_FAST | BUTTON_RIGHT: // SEEK RIGHT
+		if (ctx->displayIndex < 42 - 12) {
```
Comment (LeoHChen) [inline src/getPublicKey.c]: any definition on the 42 and 12?

Comment (LeoHChen) [reply src/getPublicKey.c]: I saw many 12 in the code, better have some explanation.

File: src/ux.h [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    uint8_t toShardStr[13];
+    uint8_t hash[32];
+	uint8_t fullStr[128]; // variable length
+    uint8_t typeStr[40]; // variable-length
+	bool initialized; // protects against certain attacks
+} signTxnContext_t;
+
+// To save memory, we store all the context types in a single global union,
+// taking advantage of the fact that only one command is executed at a time.
+typedef union {
```
Comment (LeoHChen) [inline src/ux.h]: minor, alignment.

File: src/ux.h [hunk 14]
```diff
@@ -0,0 +1,86 @@
+// Each command has some state associated with it that sticks around for the
```
Comment (LeoHChen) [inline src/ux.h]: minor, some alignment issue in the code

File: src/rlp.c [hunk 15]
```diff
@@ -0,0 +1,439 @@
+/*******************************************************************************
```
Comment (LeoHChen) [inline src/rlp.c]: is there any unit test code can be added for rlp and uint256?

File: src/harmony.c [hunk 16]
```diff
@@ -0,0 +1,118 @@
+#include <stdbool.h>
```
Comment (LeoHChen) [inline src/harmony.c]: any copyright header? like BSD

File: Dockerfile [hunk 17]
```diff
@@ -1,21 +1,6 @@
...
-&& tar -jxf make-4.2.1.tar.bz2 \
-&& cd make-4.2.1 \
-&& ./configure --prefix=/usr \
-&& make \
-make install
-
-
-USER test
+docker pull nbasim/ledger-blue-sdk
+docker run -t -i nbasim/ledger-blue-sdk /bin/bash
```
Comment (fxfactorial) [inline Dockerfile]: Why using docker in the dockerfile?

**Pull request merged**